src/Compilers/Core/Portable/Compilation/IExpression.cs

```diff
@@ -0,0 +1,778 @@
+        StringLike,
+
+        DelegateEqual,
+        DelegateNotEqual,
+
+        NullableEqual,
+        NullableNotEqual,
+
+        ObjectLess,
+        ObjectLessEqual,
+        ObjectEqual,
+        ObjectNotEqual,
+        ObjectVBEqual,
+        ObjectVBNotEqual,
+        ObjectLike,
+        ObjectGreaterEqual,
+        ObjectGreater,
+
+        EnumEqual,
+        EnumNotEqual,
+        EnumLess,
+        EnumLessEqual,
+        EnumGreaterEqual,
+        EnumGreater,
+
+        PointerEqual,
+        PointerNotEqual,
+        PointerLess,
+        PointerLessEqual,
+        PointerGreaterEqual,
+        PointerGreater,
+
+        DynamicEqual,
+        DynamicNotEqual,
+        DynamicLess,
+        DynamicLessEqual,
+        DynamicGreaterEqual,
+        DynamicGreater
+    }
+
+    /// <summary>
+    /// Represents a conversion operation.
+    /// </summary>
+    public interface IConversion : IWithOperator
+    {
+        IExpression Operand { get; }
+        ConversionKind Conversion { get; }
+        /// <summary>
+        /// True if and only if the conversion is indicated explicity by a cast operation in the source code.
+        /// </summary>
+        bool IsExplicit { get; }
+    }
+
+    /// <summary>
+    /// Kinds of conversions.
+    /// </summary>
+    public enum ConversionKind
+    {
+        None,
+        /// <summary>
+        /// Conversion is defined by the underlying type system and throws an exception if it fails.
+        /// </summary>
+        Cast,
+        /// <summary>
+        /// Conversion is defined by the underlying type system and produces a null result if it fails.
+        /// </summary>
+        AsCast,
+        /// <summary>
+        /// Conversion has VB-specific semantics.
+        /// </summary>
+        Basic,
+        /// <summary>
+        /// Conversion has C#-specific semantics.
+        /// </summary>
+        CSharp,
+        /// <summary>
+        /// Conversion is implemented by a conversion operator method.
+        /// </summary>
+        Operator
+    }
+
+    public interface IConditionalChoice : IExpression
+    {
+        IExpression Condition { get; }
+        IExpression IfTrue { get; }
+        IExpression IfFalse { get; }
+    }
+
+    public interface INullCoalescing : IExpression
+    {
+        IExpression Primary { get; }
+        IExpression Secondary { get; }
+    }
+
+    public interface IIs : IExpression
+    {
+        IExpression Operand { get; }
+        ITypeSymbol IsType { get; }
+    }
+
+    public interface ITypeOperation : IExpression
+    {
+        TypeOperationKind TypeOperationClass { get; }
+        ITypeSymbol TypeOperand { get; }
+    }
+
+    /// <summary>
+    /// Kinds of type operations.
+    /// </summary>
+    public enum TypeOperationKind
+    {
+        None,
+
+        SizeOf,
+        TypeOf
+    }
+
+    public interface ILambda : IExpression
+    {
+        IMethodSymbol Signature { get; }
+        IBlock Body { get; }
```

Comment:
So this looks like we lowered a bit.  Because if i have: a => b, then i'm surprised that we would have a block body.  Would that block body then have a single 'return' operation?  I think a better way would be to have an IOperation body here which could be an expression or block.